crates/exex/src/manager.rs

```diff
@@ -0,0 +1,446 @@
+            while let Poll::Ready(Some(event)) = exex.receiver.poll_recv(cx) {
+                debug!(?event, "received event from exex {}", exex.id);
+                exex.metrics.events_sent_total.increment(1);
+                match event {
+                    ExExEvent::FinishedHeight(height) => exex.finished_height = Some(height),
+                }
+            }
+        }
+
+        // update watch channel block number
+        let finished_height = self.exex_handles.iter_mut().try_fold(u64::MAX, |curr, exex| {
+            let height = match exex.finished_height {
+                None => return Err(()),
+                Some(height) => height,
+            };
+
+            if height < curr {
+                Ok(height)
+            } else {
+                Ok(curr)
+            }
+        });
+        if let Ok(finished_height) = finished_height {
+            let _ = self.finished_height.send(Some(finished_height));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// A handle to communicate with the [`ExExManager`].
+#[derive(Debug)]
+pub struct ExExManagerHandle {
+    exex_tx: UnboundedSender<CanonStateNotification>,
+    num_exexs: usize,
+    is_ready_receiver: watch::Receiver<bool>,
+    is_ready: WatchStream<bool>,
+    current_capacity: Arc<AtomicUsize>,
+    finished_height: watch::Receiver<Option<BlockNumber>>,
+}
+
+impl ExExManagerHandle {
+    /// Synchronously send a notification over the channel to all execution extensions.
+    ///
+    /// Senders should call [`Self::has_capacity`] first.
+    pub fn send(
+        &self,
+        notification: CanonStateNotification,
+    ) -> Result<(), SendError<CanonStateNotification>> {
+        self.exex_tx.send(notification)
+    }
+
+    /// Asynchronously send a notification over the channel to all execution extensions.
+    ///
+    /// The returned future resolves when the notification has been delivered. If there is no
+    /// capacity in the channel, the future will wait.
+    pub async fn send_async(
+        &mut self,
+        notification: CanonStateNotification,
+    ) -> Result<(), SendError<CanonStateNotification>> {
+        self.ready().await;
+        self.exex_tx.send(notification)
+    }
+
+    /// Get the current capacity of the ExEx manager's internal notification buffer.
+    pub fn capacity(&self) -> usize {
+        self.current_capacity.load(Ordering::Relaxed)
+    }
+
+    /// Whether there is capacity in the ExEx manager's internal notification buffer.
+    ///
+    /// If this returns `false`, the owner of the handle should **NOT** send new notifications over
+    /// the channel until the manager is ready again, as this can lead to unbounded memory growth.
+    pub fn has_capacity(&self) -> bool {
+        self.current_capacity.load(Ordering::Relaxed) > 0
+    }
+
+    /// Returns `true` if there are ExEx's installed in the node.
+    pub fn has_exexs(&self) -> bool {
+        self.num_exexs > 0
+    }
```

Comment:
are we using it anywhere? if not, then we seem to not need `num_exexs` as well

Reply:
yes, this will be important for the execution stage - no exex's, no need to build `CanonStateNotification` (which has some overhead in terms of allocations)